module/Finna/src/Finna/RecordDriver/SolrEad3.php

```diff
@@ -893,59 +894,88 @@ public function getExtendedAccessRestrictions()
-                    $type = (string)$attr->encodinganalog;
-                    if (in_array($type, self::ACCESS_RESTRICT_TYPES)) {
-                        $type = str_replace(':', '_', $type);
-                        switch ($type) {
-                        case 'ahaa_KR7':
-                            $label = $this->getDisplayLabel(
-                                $access->p->name, 'part', true
-                            );
-                            break;
-                        case 'ahaa_KR9':
-                            $label = [(string)($access->p->date ?? '')];
-                            break;
-                        default:
-                            $label = $this->getDisplayLabel($access, 'p');
-                        }
-                        if ($label) {
-                            $restrictions[$type] = $label;
+
+        $processNode = function ($access) use (&$restrictions) {
+            $attr = $access->attributes();
+            if (! isset($attr->encodinganalog)) {
+                $restriction['general'] = array_merge(
+                    $restrictions['general'],
+                    $this->getDisplayLabel($access, 'p', true)
+                );
+            } else {
+                $type = (string)$attr->encodinganalog;
+                if (in_array($type, self::ACCESS_RESTRICT_TYPES)) {
+                    switch ($type) {
+                    case 'ahaa:KR7':
+                        $label = $this->getDisplayLabel(
+                            $access->p->name, 'part', true
+                        );
+                        break;
+                    case 'ahaa:KR9':
+                        $label = [(string)($access->p->date ?? '')];
+                        break;
+                    default:
+                        $label = $this->getDisplayLabel($access, 'p');
+                    }
+                    if ($label) {
+                        // These are displayed under the same heading
+                        if (in_array($type, ['ahaa:KR2', 'ahaa:KR3'])) {
+                            $type = 'ahaa:KR1';
                         }
+                        $restrictions[$type]
+                            = array_merge($restrictions[$type], $label);
                     }
                 }
             }
+        };
+
+        foreach ($xml->accessrestrict ?? [] as $accessNode) {
+            $processNode($accessNode);
+            foreach ($accessNode->accessrestrict ?? [] as $accessNode) {
+                $processNode($accessNode);
+                foreach ($accessNode->accessrestrict ?? [] as $accessNode) {
+                    $processNode($accessNode);
+                }
+            }
         }
 
-        // Sort and discard empty
         $result = [];
-        foreach ($restrictions as $type => $values) {
-            if (empty($values)) {
-                unset($restrictions[$type]);
+
+        // Sort
+        $order = self::ACCESS_RESTRICT_TYPES;
```

Comment:
Sorttaus olisi tehokkaampi, jos ihan ensin `$order` tehtäisiin array_flip:llä. Sitten olis simppelimmin `$pos1 = $order[$a] ?? 9999;` tms. Ja sitten voi palauttaa suoraan esim. `return $pos1 - $pos2;`.

Reply:
Totta, muutin näin.